### ISSUE TITLE
Add support for `aioboto3`

### ISSUE THREAD

**Issue by takeda** [youtype/mypy_boto3_builder#153]
**Describe the bug**
No types for responses when using `types-aiobotocore`.

**To Reproduce**

1. Install `types-aiobotocore[ec2]`
2. Run `mypy`/`pyright` on the following code sample and `--strict` option.

```python
    session = aioboto3.Session()
    async with session.client("ec2", region_name=destination_region) as ec2:
          response = await ec2.copy_image(
              Name=original_name, SourceRegion=source_ami.region, SourceImageId=source_ami.id, DryRun=config.dry_run
          )
          destination_ami = response["ImageId"]
```

The last line returns: `error: Returning Any from function declared to return "str"` and requires `cast()` this works fine with boto3-stubs.

**Additional context**
Your OS, `boto3-stubs` installation method, `boto3` version, etc.
Darwin, `types-aiobotocore` 2.3.0 installed via `poetry`

**Comment by vemel** [comment]
`types-aiobotocore` does not support implicit typing for `aioboto3`, only for raw `aiobotocore`. You can use explicit type annotations:

```python
import aioboto3
from types_aiobotocore_ec2.client import EC2Client

session = aioboto3.Session()

ec2: EC2Client  # set type explicitly

async with session.client("ec2", region_name=destination_region) as ec2:
    response = await ec2.copy_image(
        Name=original_name,
        SourceRegion=source_ami.region,
        SourceImageId=source_ami.id,
        DryRun=config.dry_run,
    )
    destination_ami = response["ImageId"]
```

Full documentation: https://youtype.github.io/types_aiobotocore_docs/types_aiobotocore_ec2/client/

Let me know if this solves your issue.

**Comment by vemel** [comment]
Ideally, I should add `types-aioboto3` package. I think it is doable in a couple of weeks.

**Comment by takeda** [comment]
I got used to how nice using standard `boto3` is with `boto3-stubs`. But if you could do that, that would be awesome. Thank you.

**Comment by vemel** [comment]
@takeda `types-aioboto3` released and works the same as `boto3-stubs`, so you can install it with

```bash
python -m pip install 'types-aioboto3[ec2]'
```

Please let me know if it works as it should for you.

**Comment by takeda** [comment]
I still have to list `types-aiobotocore[ec2]` in dependencies, right?

**Comment by takeda** [comment]
It appears to work. And `cast()` and `# type: ignore` is no longer needed. Thank you!

**Comment by MiloLug** [comment]
Ok, but how to specify types explicitly? I mean, I can use `types_aiobotocore_xxx`, but it isn't fully compatible with aioboto3.

For example, with the `Bucket` resource, it can't represent the fields correctly - they are `Awaitable[]` in places, where aioboto3 provides just simple value.

Here is working code, but PyCharm doesn't think so due to aiobotocore's Bucket.

![image](https://user-images.githubusercontent.com/33329419/215079289-60393ece-c520-438c-a1bf-5eb74d6060f3.png)

**Comment by MiloLug** [comment]
Any news?

**Comment by vemel** [comment]
I still did not figure out how `aioboto3` differentiates Awaitable fields. Any help would be appreciated.

The only idea I have for now is to actually create a client and `inspect` all fields and methods.

**Comment by vemel** [comment]
@MiloLug Finally, it is fixed in `types-aiobotocore 2.4.2.post2`. Please test it and let me know if it works as it should.

I checked `aioboto3` code and found out that only non-identifier and non-reference attributes are `Awaitable`.

**Comment by MiloLug** [comment]
Sorry I couldn't reply earlier. Now it works almost as expected. However I've spotted some problems just now.

Stubs for aiobotocore use `__iter__` instead of `__aiter__`.

Also filtering functions (`all`, `filter`, `limit`, `page_size`, `pages`) of the resource collections seem to be broken.
As far as I can see from aioboto3 `AIOResourceCollection`, they override only `pages` to be async, **but** you can't make it async either in the stubs since it should return `AsyncIterator`, and this isn't possible to achieve without specifying `yield` in a function. 

So the stubs should be (object versions collection as an example):
```python
async def delete(self, ...) -> List[DeleteObjectsOutputTypeDef]: ...
def filter(self, ...) -> "BucketObjectVersionsCollection": ...
def limit(self, count: int) -> "BucketObjectVersionsCollection": ...
def page_size(self, count: int) -> "BucketObjectVersionsCollection": ...
def pages(self) -> AsyncIterator[List["ObjectVersion"]]: ...
def __iter__(self) -> AsyncIterator["ObjectVersion"]: ...
```

I'm not sure how you can distinguish the methods that should be async (as `delete`). But I think, if you just make all methods of the base `AIOResourceCollection` synchronous and replace Iterators with AsyncIterators, that will work.

I'm truly sorry I can't contribute, since I don't have much time and didn't research your module that enough. But I can try this weekend

**Comment by vemel** [comment]
@MiloLug thank you for the detailed research on the issue.

I changed the logic for `AIOResourceCollection` to the following:
- all common methods that return self (`filter`, `limit`, `page_size`, `all`) are synchronous
- `pages` sync method return `AsyncIterator[List[CollectionItem]]`
-`__iter__` method returns `NoReturn` because it always raises `NotImplementedError`
- added `__aiter__` sync method that returns `AsyncIterator[CollectionItem]`
- all Collection-specific methods (like `delete`) are asynchronous

This change will be included in the next release today or tomorrow. Feel free to review https://github.com/youtype/mypy_boto3_builder/commit/3bf83d096296dbbf58adf33abe1adc3b8d252457

**Comment by vemel** [comment]
@MiloLug  The changes were included to `types-aioboto3 11.0.1.post1` and underlying `types-aiobotocore 2.4.2.post3`. Please test the new release and let me know if it works as expected.

**Comment by MiloLug** [comment]
Works like a charm
